valkyrietry.go

```diff
@@ -0,0 +1,253 @@
+		}
+
+		err := f()
+
+		if err == nil {
+			return nil
+		}
+
+		retryInterval := valkyrietry.getRetryIntervalValue(currentInterval)
+		currentInterval = time.Duration(float32(currentInterval) * valkyrietry.Configuration.RetryBackoffMultiplier)
+
+		currentAttempt++
+
+		timer.Start(retryInterval)
+
+		select {
+		case <-valkyrietry.ctx.Done():
+			return valkyrietry.ctx.Err()
+		case <-timer.C():
+		}
+	}
+}
+
+// DoWithData
+// Start the retry mechanism with any given data to receive and run the process until the `RetryFunc`
+// successfully returns with the data (without error) or until the maximum number of retry attempts is exceeded.
+//
+// This function ensures that the given `RetryFunc` will run at least once.
+func DoWithData[T any](f RetryFuncWithData[T], options ...Option) (T, error) {
+	valkyrietry := defaultValkyrietry(options...)
+
+	currentAttempt := 0
+	currentInterval := valkyrietry.Configuration.InitialRetryDelay
+
+	timer := NewTimer()
+
+	defer func() {
+		timer.Stop()
+	}()
+
+	var response T
+
+	for {
+		if currentAttempt > int(valkyrietry.Configuration.MaxRetryAttempts) {
+			return response, ErrMaxRetryAttemptsExceeded
+		}
+
+		response, err := f()
+
+		if err == nil {
+			return response, nil
+		}
+
+		retryInterval := valkyrietry.getRetryIntervalValue(currentInterval)
+		currentInterval = time.Duration(float32(currentInterval) * valkyrietry.Configuration.RetryBackoffMultiplier)
+
+		currentAttempt++
+
+		timer.Start(retryInterval)
+
+		select {
+		case <-valkyrietry.ctx.Done():
+			return response, valkyrietry.ctx.Err()
+		case <-timer.C():
+		}
+	}
+}
+
+// DoWithDataAndContext
+// Start the retry mechanism with any given data to receive and a context, and continue running the process until the `RetryFunc`
+// successfully returns with the data (without error) or until the maximum number of retry attempts is exceeded.
+//
+// This function ensures that the given `RetryFunc` will run at least once.
+func DoWithDataAndContext[T any](ctx context.Context, f RetryFuncWithData[T], options ...Option) (T, error) {
+	valkyrietry := defaultValkyrietryWithContext(ctx, options...)
+
+	currentAttempt := 0
+	currentInterval := valkyrietry.Configuration.InitialRetryDelay
+
+	timer := NewTimer()
+
+	defer func() {
+		timer.Stop()
+	}()
+
+	var response T
+
+	for {
+		if currentAttempt > int(valkyrietry.Configuration.MaxRetryAttempts) {
+			return response, ErrMaxRetryAttemptsExceeded
+		}
+
+		response, err := f()
+
+		if err == nil {
+			return response, nil
+		}
+
+		retryInterval := valkyrietry.getRetryIntervalValue(currentInterval)
+		currentInterval = time.Duration(float32(currentInterval) * valkyrietry.Configuration.RetryBackoffMultiplier)
+
+		currentAttempt++
+
+		timer.Start(retryInterval)
+
+		select {
+		case <-valkyrietry.ctx.Done():
+			return response, valkyrietry.ctx.Err()
+		case <-timer.C():
+		}
+	}
+}
+
+func (v *Valkyrietry) getRetryIntervalValue(currentInterval time.Duration) time.Duration {
```

Comment:
**thought** : This function is to model how often the *library* will do a retry for function.

**suggestion (non-blocking)** : I wonder whether it's possible to add custom function to model the retries mechanism, I expect it would have the form of : 
`func(lastRetryAt time.Time, lastDuration time.Duration) time.Duration`. With this, we could model custom delay like exponential duration or other flow control mechanism. (open for discussion).

Reply:
Interesting!

Could we possibly handle this in a separate PR dedicated to another feature? What do you think?

Reply:
> Could we possibly handle this in a separate PR dedicated to another feature? What do you think?

Agree, we could implement it as another Option func (and propagate default behavior into another Option func) in the next MR as improvement.